File: Kattis/kattis_palindromic.py
def main():

	cases = int( input() )

	for i in range( cases ):

		name   = input()
		length = len( name )
		table  = list()

		for i in range( length ):

			table.append( [0] * length )
			table[i][i] = 1

			if i != length - 1:

				table[i][i+1] = 2

				if name[i] == name[i+1]:

					table[i][i+1] += 1

		for k in range( 2, length ):
			for i in range( length - k ):

				j = k + i

				table[i][j] = ( table[i+1][j] + table[i][j-1] - \
								table[i+1][j-1] ) % ( 10e9 + 7 )

				if name[i] == name[j]:

					table[i][j] += ( table[i+1][j-1] + 1 ) % ( 10e9 + 7 )

		print( int( table[0][-1] % ( 10e9 + 7 ) ) )

if __name__ == "__main__":

	main()

File: Kattis/kattis_palindromic.cpp
#include <iostream>
#include <cstring>

using namespace std;

long long table[2010][2010];

int main() {

    int cases, j;

    cin >> cases;

    while ( cases-- ) {

        string name;

        memset( table, 0, sizeof( table ) );

        cin >> name;

        for ( int i = 0; i < name.length(); i++ ) {

            table[i][i] = 1;
            table[i][i+1] = 2 + ( name[i] == name[i+1] );
        }

        for ( int k = 2; k < name.length(); k++ ) {

            for ( int i = 0; i < name.length() - k; i++ ) {

                j = k + i;

                table[i][j] = ( table[i+1][j] + table[i][j-1] - table[i+1][j-1]
                                + ( name[i] == name[j] ) * ( table[i+1][j-1] +
                                1 ) ) % 1000000007;
            }
        }

        cout << table[0][name.length() - 1] << endl;
    }

    return 0;
}
